### PR TITLE
Refactor options and files resolve functionality to allow Go invocations

### PR DESCRIPTION
**Pull request by albertomilan**
The purpose of this change is to allow other Go projects to make use of the `ko resolve` feature programmatically. In order to do so, we've moved all the model options to a common module (`parameters`) and kept all Cobra parsing and CLI related code in `commands`.
Besides this, we have made public the MakeBuilder, MakePublisher and ResolveFilesToWriter functions. These will be used by other projects in order to perform the desired functionality.

This change does not introduce any improvements or regression beyond the new structuring and layout.

### CODE REVIEW

File: pkg/build/gobuild.go [hunk 1]
```diff
@@ -443,11 +443,11 @@ func (g *gobuild) tarKoData(ref reference) (*bytes.Buffer, error) {
 }
 
 // Build implements build.Interface
-func (gb *gobuild) Build(ctx context.Context, s string) (v1.Image, error) {
+func (g *gobuild) Build(ctx context.Context, s string) (v1.Image, error) {
```
[the remark this answers is not in the document]
Comment (albertomilan) [reply pkg/build/gobuild.go:446]: I agree. This was a non-functional code consistency update. I will remove it for the next update

**Comment by albertomilan** [comment]
The purpose of this change is to make use of the exact same functionality `ko resolve` provides from other existing Go packages. By doing so, we will be able to introduce better testing practices (mocks).
I am not sure about the initialization performed for default values you mention. We want to mimic the **exact same** functionality but programatically. There's a chance we can obtain this same functionality by just importing the commands. I'm fairly new to Go, so any advice is more than welcome!

**Comment by ImJasonH** [comment]
Do you really want to support _every_ option of `ko resolve`? Options like `--watch` add quite a bit of complexity that you might not need, and things like publishing to local or insecure registries might also not be necessary based on your use case. Even things like resolving multiple files might be something you can simplify if you don't think you'll need to resolve it.

I'd love to get to a point where you can use some exported method in `pkg/resolve`, maybe even [`resolve.ImageReferences`](https://godoc.org/github.com/google/ko/pkg/resolve#ImageReferences), with a [real go builder](https://godoc.org/github.com/google/ko/pkg/build#NewGo) and a [real publisher](https://godoc.org/github.com/google/ko/pkg/publish#NewDefault). If there's something that method doesn't support, we should consider adding it.

**Comment by develar** [comment]
I want to use `ko publish` in my Pulumi code. As far I see, this PR was not accepted, right? And function `gobuildOptions` or `makeBuilder` are not exported.

**Comment by jonjohnsonjr** [comment]
Yeah it seems the PR was closed. I have a couple mostly cosmetic issues with it, but I'm not opposed to exposing something like this. Most of this code is structured around supporting the CLI, so we'd need to do some refactoring to make it "make sense" as a library. I agree with Jason that watch mode might not make sense for a library, but I could be persuaded otherwise.

**Comment by albertomilan** [comment]
Yes, this PR was closed as it was introducing direct dependencies and complicated invocation patterns into our code base. Unless `ko` is refactored more generally as a library and invocation is simplified, I would suggest to avoid light attempts. It will introduce code duplicity in your modules and brittle logic.

**Pull request closed without merging**